### ISSUE TITLE
Missing smithy-trait-codegen union support

### ISSUE THREAD

**Issue by msosnicki** [smithy-lang/smithy#2268]
We're using the new smithy-trait-codegen module to generate the java trait files for the protocols. Seem that the unions are not supported yet https://github.com/smithy-lang/smithy/blob/1805e515bc3877a02aa8351da1a7ea62e92a2c8f/smithy-trait-codegen/src/main/java/software/amazon/smithy/traitcodegen/TraitCodegenSymbolProvider.java#L183. 

Is it something on the radar for the upcoming releases? If not, would contribution speed up the process?

Thanks

**Comment by hpmellema** [comment]
We do have plans to add support for union traits in a future release.

**Comment by joelvanderhoof** [comment]
> We do have plans to add support for union traits in a future release.

Will this include support for untagged and discriminated unions?

**Comment by hpmellema** [comment]
The support for union traits mentioned here is for generating the java implementation of custom Smithy Union traits such as:
```
@trait 
union MyUnion {
    option1: String 
    option2: Integer
}
```

I was not referring to adding additional behavior traits to modify the behavior of Smithy unions. Support for union trait generation will not include anything relating to untagged unions.

**Comment by msosnicki** [comment]
@hpmellema Will it also support the following case?
```smithy
@trait
structure MyTrait {
    union: MyUnion
} 
union MyUnion {
    option1: String 
    option2: Integer
}
```

During decoding, it would work only for tagged unions, right? For example:
```smithy
@MyTrait(
    union: {
        option1: "Foo"
    }
}
structure StructWithTrait {}

```

**Comment by hpmellema** [comment]
Yes, we plan to support nested unions within shape definitions, and yes, on node deserialization only tagged unions would be supported.
